lambda-buffers-codegen/src/LambdaBuffers/Codegen/Rust/Print/LamVal.hs

```diff
@@ -0,0 +1,398 @@
+import Data.Map qualified as Map
+import Data.Map.Ordered qualified as OMap
+import Data.ProtoLens (Message (defMessage))
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Traversable (for)
+import LambdaBuffers.Codegen.LamVal qualified as LV
+import LambdaBuffers.Codegen.LamVal.MonadPrint qualified as LV
+import LambdaBuffers.Codegen.Rust.Print.Syntax qualified as R
+import LambdaBuffers.Codegen.Rust.Print.TyDef qualified as TD
+import LambdaBuffers.Compiler.LamTy qualified as LT
+import LambdaBuffers.ProtoCompat qualified as PC
+import Prettyprinter (Doc, Pretty (pretty), align, angles, braces, brackets, colon, comma, dot, dquotes, encloseSep, equals, group, langle, lbracket, line, lparen, parens, pipe, punctuate, rangle, rbracket, rparen, semi, space, vsep, (<+>))
+import Proto.Codegen_Fields qualified as P
+
+caseIntERef :: R.QValName
+caseIntERef = R.Qualified'LibRef (R.MkCrateName "lbr-prelude") (R.MkModuleName "lamval") (R.MkValueName "case_int")
+
+bigInt :: R.QValName
+bigInt = R.Qualified'LibRef (R.MkCrateName "num-bigint") (R.MkModuleName "") (R.MkValueName "BigInt")
+
+vecAsSlice :: R.QValName
+vecAsSlice = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "vec") (R.MkValueName "Vec::as_slice")
+
+vecMacro :: R.QValName
+vecMacro = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "") (R.MkValueName "vec!")
+
+fromU32Trait :: R.QTraitName
+fromU32Trait = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "convert") (R.MkTraitName "From<u32>")
+
+fromStrTrait :: R.QTraitName
+fromStrTrait = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "convert") (R.MkTraitName "From<&str>")
+
+cloneTrait :: R.QTraitName
+cloneTrait = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "clone") (R.MkTraitName "Clone")
+
+boxNew :: R.QValName
+boxNew = R.Qualified'LibRef (R.MkCrateName "std") (R.MkModuleName "boxed") (R.MkValueName "Box::new")
+
+phantomData :: R.QTyName
+phantomData = R.qLibRef R.MkTyName "std" "marker" "PhantomData"
+
+{- | Clone a value (converting a type to owned)
+ As in codegen we cannot know whether a type is owned or borrowed, we make sure to have an owned type by making a clone.
+ We must also borrow it first for the same reason (worst case it's a double && which Rust knows how to deal with)
+-}
+clone :: Doc ann -> Doc ann
+clone = useTraitMethod cloneTrait "clone" . borrow
+
+borrow :: Doc ann -> Doc ann
+borrow doc = "&" <> doc
+
+fromU32 :: Doc ann -> Doc ann
+fromU32 = useTraitMethod fromU32Trait "from"
+
+fromStr :: Doc ann -> Doc ann
+fromStr = useTraitMethod fromStrTrait "from"
+
+useTraitMethod :: R.QTraitName -> Text -> Doc ann -> Doc ann
+useTraitMethod trait method d = angles ("_" <+> "as" <+> R.printRsQTraitName trait) <> R.doubleColon <> pretty method <> parens d
+
+throwInternalError :: MonadPrint m => String -> m a
+throwInternalError msg = throwError $ defMessage & P.msg .~ "[LambdaBuffers.Codegen.Rust.Print.LamVal] " <> Text.pack msg
```

Comment:
Oh man the error handling is so bad :D I think it would have been better is msg is a list of msgs. Oh well